Add frontend end to end testing

tallcoleman opened the pull request:
Added an end to end test for the `/submissions` route using Playwright. 

Notable features:

- Added playwright to frontend for end to end testing, created an end-to-end test for the `/submissions` route, and added it to the CI.
- When a running database is required for tests or a dev server, the make targets now use Docker to run it in a container automatically. Installing and configuring Postgres on your development machine is no longer required.
- Tests that require the database can be run at the same time as development servers.
- Added a make target `test-all` to run all tests at once (api, frontend, and end-to-end).

Notable changes:

- Added a pytest marker for API test that requires a running database.
- manage.py recreate_db function now creates `bikespace_dev` or `bikespace_test` if they don't already exist - removes the need to manually create the database instance or script creating it.
- manage.py new test_db_server function to ensure that a script waits for a postgres server to be running before it continues.
- Added tracking of hydration state to `SelectInput` component to prevent interaction before hydration.
- Added id to `Summary` and `Time` components to help with testing and connecting form labels to form inputs.
- Added and updated relevant documentation.

tallcoleman committed:
Set up playwright

tallcoleman committed:
Added e2e test for submission form

- added baseURL and env variables to config
- disabled inputs on submission issue page to prevent interaction before hydration

tallcoleman committed:
Updated notes and submission test

Added network isolation to submission test

tallcoleman committed:
Added e2e testing to makefile and CI

tallcoleman committed:
Updated CI test and mocked API

- Removed live API from the e2e test for now to reduce complexity of set-up
- Removed old autogenerated CI test not being used

tallcoleman committed:
Update test-preview.yml

- fix path to playwright reports
- make e2e testing prereq for cloudflare publish

tallcoleman committed:
Submission test fixes

- tried to improve reliability of first link click on webkit
- added disable before render to all first question inputs and made disabled value optional for other uses

tallcoleman commented:
Things to consider before completing draft:

- [x] re-add use of local API server during testing (and consider how to make the set-up as easy as possible)
  - [x] have gotten it working with frontend, db, and api all running in docker
  - [x] ~~(not possible) Change test frontend container to use build instead of dev server to prevent actions flaking out first time~~
  - [x] Set up so that starting docker containers, running tests, and then cleaning up all happens in one command
- [x] add some documentation/examples
- [x] ~~(unsure how to do this) should the end to end tests be considered for code coverage? Does it integrate in some way with the coverage results generated by pytest or jest?~~
- [x] Streamline the CI action and clean up any rough edges

tallcoleman committed:
Merge branch 'main' into frontend-end-to-end-testing

tallcoleman committed:
Submission API route should pull from env variable

tallcoleman committed:
Fix schema used for api submissions POST

tallcoleman committed:
Update submission route to align with e2e testing

- need to keep id for submission summary div
- getByLabelText preferred over using TestId (better simulation of user behaviour and providing an input label improves accessibility)

tallcoleman committed:
Merge branch 'main' into frontend-end-to-end-testing

tallcoleman committed:
Update package-lock.json

Ran npx update-browserslist-db@latest

tallcoleman committed:
Update submission.spec.ts

reduce flakiness of first link click

tallcoleman committed:
WIP - get API running in docker

Run with `docker compose -f compose-test.yaml up` / down

tallcoleman committed:
Update .dockerignore

tallcoleman committed:
Add in frontend to docker-compose

tallcoleman committed:
Confirmed playwright runs with app docker set-up

tallcoleman committed:
Make separate dockerfile for dev frontend

- also tested using more/longer toPass intervals on flaky part of e2e testing with mixed results

tallcoleman committed:
Revert previous changes

Working branch with notes saved to experiment-dockerize-app branch

tallcoleman committed:
Clean up imports

tallcoleman committed:
Set up one command api with db launch

End to end testing can now be run with just the one command e2e-frontend
- added manage.py command to test database connection
- added failsafe to create database in postgres if it doesn't exist
- configured one-liner to start database using Docker and run flask app for testing
- added one-liner to playwright database requirements

tallcoleman committed:
CI fix - remove launch-db from run-pytest

tallcoleman committed:
Added more elegant skip of docker run on CI

tallcoleman committed:
Add postgres service to e2e test action

- also removed database set up step for the api tests that is no longer needed

tallcoleman committed:
Added more informative comment to API submission used in e2e test

tallcoleman committed:
Fixed interaction before hydration and hydration interrupt issues for e2e test

tallcoleman committed:
Added Makefile target to run all tests

tallcoleman committed:
Bugfix - disabled property on SelectInput should be optional

tallcoleman committed:
Added test marker for pytest tests that require database

tallcoleman committed:
Added marker definition for pytest

tallcoleman committed:
Isolated test actions from dev actions to allow concurrent use

Now any of the dev servers can be run while also running a test at the same time
- servers run for e2e testing now at different ports than dev server
- running a test that hits the database will not restart the db docker container

tallcoleman committed:
Updated API readme

Removed older instructions about installing Postgres that are no longer neccessary

tallcoleman committed:
Playwright documentation updates

tallcoleman committed:
Remove import not being used

Was likely from an incorrect choice when merging from upstream

tallcoleman committed:
Remove script line no longer needed to create database

recreate-db script used in make run-pytest will create the database

tallcoleman commented:
@sonalranjit @xinweny this is now ready for review. Key details:

> Added an end to end test for the `/submissions` route using Playwright.
> 
> Notable features:
> 
> * Added playwright to frontend for end to end testing, created an end-to-end test for the `/submissions` route, and added it to the CI.
> * When a running database is required for tests or a dev server, the make targets now use Docker to run it in a container automatically. Installing and configuring Postgres on your development machine is no longer required.
> * Tests that require the database can be run at the same time as development servers.
> * Added a make target `test-all` to run all tests at once (api, frontend, and end-to-end).
> 
> Notable changes:
> 
> * Added a pytest marker for API test that requires a running database.
> * manage.py recreate_db function now creates `bikespace_dev` or `bikespace_test` if they don't already exist - removes the need to manually create the database instance or script creating it.
> * manage.py new test_db_server function to ensure that a script waits for a postgres server to be running before it continues.
> * Added tracking of hydration state to `SelectInput` component to prevent interaction before hydration.
> * Added id to `Summary` and `Time` components to help with testing and connecting form labels to form inputs.
> * Added and updated relevant documentation.

sonalranjit commented on bikespace_api/README.md:
```diff
@@ -27,52 +18,46 @@ Running the backend service:
-- In a terminal, run `$ make run-flask-app` and navigate to the local url it runs on (usually http://127.0.0.1:8000) to open the Swagger UI page.
-- On the Swagger page, you can open GET and select "Try it out" - if everything is working correctly, the API should return the dummy data.
-- If you run the frontend application (`$ make run-frontend` in a separate terminal) and use it to make a new submission, you should be able to see it by making another GET request to the API or re-executing the table view in pgAdmin.
+If `postgres` or `pg_ctl` don't work for you, you might need to add postgres to your `.bash_profile` or `.zshrc` file by adding a line like this (replace 17 with your version number):
```
This might not always be the case, for instance if only the Postgres.app is installed on the mac, it doesn't come out of the box with `postgres` or `pg_ctl`, one would have to have the postgres installed from brew i think to access all that. 
I use the graphical installation of postgres.app, it's pretty easy to manage the start and stop of a postgres database .
<img width="734" alt="Screenshot 2025-02-26 at 5 53 25 PM" src="https://github.com/user-attachments/assets/d9d32582-eaa0-4921-b74f-ffb7c1001274" />

tallcoleman replied:
Added a short note to cover instructions for Postgres.app users

sonalranjit commented on Makefile:
```diff
@@ -92,6 +106,24 @@ db-merge-heads:
+# Run a postgres Docker container unless in CI environment
+launch-db:
+ifneq ($(CI),true)
+	docker start db || docker run --name db \
```
Would it be more clear to label the container like `bikespace_db` ?

tallcoleman replied:
Great idea!

sonalranjit commented on Makefile:
```diff
@@ -92,6 +106,24 @@ db-merge-heads:
+endif
+
+stop-db:
+	docker stop db || true
```
Why do we need the `||` here?

tallcoleman replied:
Removed - was a vestige of when this would run every time before launch-db

sonalranjit reviewed (commented):
I might add some things to make the db container management a bit easier

tallcoleman commented:
Thanks @sonalranjit for your comments, I'll review them on the weekend!

tallcoleman committed:
Updates from review

- added more specific name for db postgres container
- removed bash or statement no longer needed
- added note in API documentation about Postgres.app

xinweny commented on bikespace_frontend/src/components/submission/issue/Issue.tsx:
I would make `useHydrated()` into a separate hook and import it wherever it's used since it appears in several files.

tallcoleman replied:
@xinweny I'm not really sure what you mean by this. Do you want to push a commit for it or suggest what the code change would be? Or we can talk about it tomorrow - let me know!

xinweny reviewed (commented):
Great work @tallcoleman with Docker integration, E2E testing as well as the documentation write-up!

tallcoleman committed:
Bugfix - set custom date filter based on user timezone

Custom date filter should always set the datetime interval based on the user's timezone. Test was previously flaking during the period when UTC is in a different calendar day than the user timezone.

tallcoleman commented:
@xinweny @sonalranjit made one more change to fix one of the Jest frontend tests that was flaking.

Basically, the custom date picker is intended to set a range between two dates, with the dates intended to be interpreted in the user's time zone. react-hook-form provides dates based on dates interpreted to be in UTC, so when UTC is in a different calendar day than Toronto (which is the timezone we test in), the test would flake.

For example, if the CI runs at 10pm ET on a Saturday, the UTC date is 3am on Sunday. So then the date filter range is off by one day and the test fails.

The fix:

- mocked the time of day in the test for the custom date range picker to ensure that it's always testing the failure conditions
- then updated the custom date picker to make sure it coerces the date value given by the form into the user's timezone

xinweny committed:
delete react mock

xinweny reviewed (approved):
🚲
